src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -573,5 +574,70 @@ internal static class ReadOnlySequence
         public static int StringToSequenceStart(int startIndex) => startIndex | StringStartMask;
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static int StringToSequenceEnd(int endIndex) => endIndex | StringEndMask;
+
+        // Borrowed from https://github.com/dotnet/coreclr/pull/20386
+
+        // Returns true iff the object has a component size;
+        // i.e., is variable length like string, array, Utf8String.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe bool ObjectHasComponentSize(object obj)
+        {
+            // CLR objects are laid out in memory as follows.
+            // [ pMethodTable || .. object data .. ]
+            //   ^-- the object reference points here
+            //
+            // The first DWORD of the method table class will have its high bit set if the
+            // method table has component size info stored somewhere. See member
+            // MethodTable:IsStringOrArray in src\vm\methodtable.h for full details.
+            //
+            // So in effect this method is the equivalent of
+            // return ((MethodTable*)(*obj))->IsStringOrArray();
+            Debug.Assert(obj != null);
+            return *(int*)GetObjectMethodTablePointer(obj) < 0;
+        }
+
+        // Given an object reference, returns its MethodTable* as an IntPtr.
+        //[Intrinsic]
+        //[NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static IntPtr GetObjectMethodTablePointer(object obj)
+        {
+            Debug.Assert(obj != null);
+            // We know that the first data field in any managed object is immediately after the
```

Comment:
It is not ok to use these implementation details outside CoreLib. They differ between CoreCLR, CoreRT or Mono.

If you want to do this, you need to move ReadOnlySequence to CoreLib.

Reply:
@ahsonkhan perhaps we can use an object pool like design instead. That might alleviate Jan's concern. I don't know offhand though whether it would result in substantial perf gains. I'll find you some time today to discuss. :)